### ISSUE TITLE
Internal error when using types returned by procedures

### ISSUE THREAD

**Issue by Rapiz1** [chapel-lang/chapel#16044]
### Steps to Reproduce

**Source Code:**

```chapel
use Sort only defaultComparator;

record treap {
    var comparator: record = defaultComparator; // line 6
}

enum orderedImpl {treap, skipList};
param _defaultImpl = orderedImpl.treap;

proc getTypeFromEnumVal(param val) type {
    if val == orderedImpl.treap then return treap();
}
proc getInstanceFromEnumVal(param val, comparator: record = defaultComparator) {
    if val == orderedImpl.treap then return new treap(comparator);
}
record outter {
    param implType = _defaultImpl;
    var instance: getTypeFromEnumVal(implType);

    proc init(comparator: record = defaultComparator,
              param implType: orderedImpl = _defaultImpl) {
      this.implType = implType;
      this.instance = getInstanceFromEnumVal(implType, comparator); 
    }
}
```

**Compile command:**
```
chpl a.chpl
a.chpl:4: internal error: AST-PRI-IVE-0235 chpl version 1.23.0 pre-release (a20083d603)

Internal errors indicate a bug in the Chapel compiler ("It's us, not you"),
and we're sorry for the hassle.  We would appreciate your reporting this bug -- 
please see https://chapel-lang.org/bugs.html for instructions.  In the meantime,
the filename + line number above may be useful in working around the issue.
```

**Associated Future Test(s):**
None

### Configuration Information

```
chpl version 1.23.0 pre-release (a20083d603)
Copyright 2020 Hewlett Packard Enterprise Development LP
Copyright 2004-2019 Cray Inc.
(See LICENSE file for more details)

CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_LLVM: none
CHPL_AUX_FILESYS: none

gcc (GCC) 10.1.0
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by cassella** [comment]
> // line 6

I think you misspelled `// line 4`.

**Comment by lydia-duncan** [comment]
This might be my fault, so I'll try to get to it soon

**Comment by bradcray** [comment]
@Rapiz1: I think on chat you said that you had variants of this program that worked fine, is that right?  Any idea what the key difference between those variants and this one is?

**Comment by Rapiz1** [comment]
> @Rapiz1: I think on chat you said that you had variants of this program that worked fine, is that right? Any idea what the key difference between those variants and this one is?

@bradcray The difference is Line 4. The other has certain fields like `int`. This one has a `record`.

**Comment by lydia-duncan** [comment]
Wait, is `record` a valid type declaration?  That obviously shouldn't give an internal compiler error, but that's also not a type, it's a type category, right?

**Comment by lydia-duncan** [comment]
I'm kinda surprised that's not a syntax error

**Comment by bradcray** [comment]
@lydia-duncan: Recall that @mppf recently made `record` into a generic type specifier that says "this will be some record, though I'm not going to say which one."  That was introduced primarily for use as a formal argument constraint I believe, but I don't see any reason it couldn't be used here as well (though I'm not certain whether it's something we've tested much / at all).

@Rapiz1: If this is blocking you, the next thing I'd try is whether removing that `: record` constraint causes it to work.

**Comment by lydia-duncan** [comment]
So there's two things I've found that happen here.  First, the most simplified version of the program here that exhibits the bug is this:
```chapel
record One {
  var x: record = new Two(2);
}
record Two {
  var y: int;
}
```
Basically, something goes wrong when we have a declaration of a type like this but we never use it.  This seems to be a problem with the `resolveAutoCopies` operation in function resolution - we run into the internal error when trying to resolve the setter for the field while resolving the type's `init=` function, as triggered by resolveAutoCopies.

Interestingly, there's a different issue when you try to use the type and write an instance of it:
```chapel
record One {
  var x: record = new Two(2);
}
record Two {
  var y: int;
}

proc main() {
  var foo: One;
  writeln(foo);
}
```
The above program fails with the message:
```
$CHPL_HOME/modules/internal/ChapelIO.chpl:217: In function 'isIoField':
$CHPL_HOME/modules/internal/ChapelIO.chpl:218: error: the type of the actual argument 'call_tmp' is generic
note: generic actual arguments are not currently supported
$CHPL_HOME/modules/internal/ChapelIO.chpl:273: Function 'isIoField' instantiated as: isIoField(x: One, param i = 1)
```
I haven't tried to hunt down why that in particular is happening, but it definitely also seems wrong.

I had versions of this program that worked without incident, though.  For instance, this program compiles and runs:
```chapel
var defaultVal: Two;

record One {
  var x: record = defaultVal;
}
record Two {
  var y: int;
}

proc whatever (param x) {
  if x == 2 then return new One(defaultVal);
}
proc main() {
  var foo: One = whatever(2);
  writeln(foo);
}
```
I think in this case, the `init=` for returning the type from the function is set up appropriately, so we don't try to set it up inappropriately when we reach `resolveAutoCopies` later in function resolution.

**Comment by lydia-duncan** [comment]
I don't think this is due to code I have written - it definitely seems like an oversight when implementing the generic `record` label.  I might still be able to solve it, but I'll probably have to touch base with @mppf on what we expect to happen in this scenario

**Comment by lydia-duncan** [comment]
I think the issue with the first case in my earlier comment is that we try to resolve an inserted call with an instance we created, but there may be something off with the temporary we generate for this case.  I think this explains why calls to the `init=` function work when called directly - in those cases, we have an actual instance that has been properly initialized, but for the temporary we created, all we have is the generic record type declaration and we know the field's default value, but we haven't done anything with it.  The particular instance that causes problems was created later in the call stack than resolveAutoCopies (fixupDefaultInitCopy and findCopyInitFn are responsible for it during candidate resolution), but the temporary we create for resolveAutoCopies is probably not much better.

**Comment by mppf** [comment]
After PR #16460, `test/types/records/generic/genericFieldDecl.chpl` is working as expected but `test/types/records/generic/genericFieldDecl-used.chpl` is not yet. (PR #16460 does provide a better error, though). There are design issues relating to how to get that test to work that are raised in issue #16508.

**Comment by mppf** [comment]
For fields declared without a type or with a generic type, it's now a compilation error to try to default-initialize them whether using a user-provided initializer or a compiler-generated one. See issue #16508 and PR #22836.

It might be sufficient for these patterns to be compilation errors. If not, the main challenge is what to name the argument to contain the type for such fields (e.g. with `record R { var x; }` and we have `var y: R(int);`; how do we call the initializer to default init? An `x` formal would normally be for specifying the value. Note that https://chapel-lang.org/docs/master/technotes/initTypeAlias.html describes how the compiler adds named arguments in such cases of default initialization with instantiated types.

Some ideas that have come up on other issues (especially #16508):
1. use a formal name of `xType`
2. use a formal name of `x.type`
3. require the initializer to use `this.type` in this scenario to get the field type
4. use a formal name of `x` but use separate initializers for the default init and for `new` calls providing some values; in particular the default initializer can use type arguments for all such fields and the value initializer can expect values be provided; so the compiler would generate `proc R.init(type x)` and `proc R.init(in x)` for the example `R`.
5. enable default initialization through the initializer like `proc R.init(in x)` by computing the default of whatever type is passed in; so `0` for integers.
